Tutorial project 3 (#7)
MS-SJF-Projects/web-game-complete: modules/03_in_memory_storage/README.md

# In memory storage

## Goal

Implement an interactive web app where you can:
- upload words to the game dictionary;
- play a game of guessing a word from the dictionary.

> ## Tutorial

Reviewer: You changed the style of this tutortial compared to other tutortials quite a lot. Can you also modify the other tutortials in this style, or converge the style of this to fit more in with the others?

Author: That's exactly why I wanted it to be reviewed sooner than later :)
And this PR in its description suggests how to change the structure of the tutorial for project 2, if we go for this style:
https://github.com/MS-SJF-Projects/web-game-complete/pull/8 

Would you agree on this structure - goal/steps as headers/recap + explaining what to do and how, but without details? 
 Do you think it works better, or would you remove certain parts?
 
I think we need a unified style, and that everybody should improve their tutorials themselves, because it looks like a major rewriting. It would be weird / not very team-player of me if I came and violated the sense of ownership by rewriting everything because I like it this or that way :)

Reviewer: I think your style looks fine, but I don't want to refactor the other tutorials because you introduced a new style 😆 

Also @LauraDamianTNA should weigh in on her opinions on the two styles :)

Reviewer 2: I agree the styles are different, and I also think they are both clear. I created a WI https://github.com/MS-SJF-Projects/web-game-complete/issues/11 to do this, but I would not make it a prio at this moment and I would focus on finishing what we proposed ourselves (adding DB, pictures, blob storage).

Now it's time to **upload** data to the server! In our case, upload words to the dictionary.

To do this, we'll need:
1. a text field on the web page with a button "submit" (this will send data to the server)
2. an endpoint on the server that would accept this data.

#### 1. Create HTML page for word upload

Let's create a web page that will contain a text field for the word to upload and a button `submit`.

Create a new file `upload_word.html` in the folder `templates`.
Inherit it from `base.html`, like we did it in the previous stage, and put the following form inside `block content`:

```html
<form action = "/upload_word" method = "POST" enctype="multipart/form-data">
    <div class="form-group">
        <label for="secretId">Secret</label>
        <input type="text" class="form-control" id="secretId" name="secretWord" aria-describedby="secretHelp" placeholder="Secret Word">
        <small id="secretHelp" class="form-text text-muted"></small>
    </div>

	<button type="submit" class="btn btn-primary">Submit</button>
</form>
```

#### 2. Serve this HTML page on HTTP endpoint

To let users open this page, add a new endpoint in `api.py`:

> ```python

Reviewer: Again, we already covered adding endpoints, do we need to give explicit instructions?

@app.route('/words', methods=['GET'])
def upload_word_page():
    return render_template('words.html')
```

Now start the app, and verify that you can see this form on the endpoint `/words`.

> Most likely, the app will start on localhost:5000. In this case you can use this link to open the page:

> > [http://localhost:5000/words](http://locahost:5000)

Reviewer: ```suggestion
> <http://localhost:5000/words>
```

When you type in a word in the form and click "submit", your browser will try to
send it to the endpoint `/upload_word`. This is the endpoint we specified in the
`action` attribute of the form.

Because this endpoint doesn't exist yet, in the browser you'll see a "Not Found" page,
and in the logs of the running application you'll see the following entry:

> ```

Reviewer: It's shell output, but you can use txt instead if you want

127.0.0.1 - - [01/Jan/2022 11:12:34] "POST /upload_word HTTP/1.1" 404 -
```

#### 3. Accept form data on the server

Let's create the endpoint `/upload_word` that will actually accept the form data.

```python
from flask import redirect, request

@app.route('/upload_word', methods=['POST'])  # browsers send POST request when submitting a form
def upload_word():
    # request.form is a special variable in Flask that will contain the form data
    secret_word = request.form['secretWord']  # note the "name" attribute of the <input> we have in HTML
    print("Uploaded word " + repr(secret_word))
    return redirect('/')  # redirect back to the main page
```

Run the application, try uploading a word, and verify that the upload works now. 
Verify that in the application log you see the uploaded word. 

#### 4. Give user feedback

> It's good to give the user know too, that the word was uploaded.

Reviewer: ```suggestion
It's good to let the user know that the word was uploaded too.
```

Flask has a concept of flashing that we can use. In the logic of our endpoint we 
would "flash" one or more messages, and these messages will become available 

> to us when rendering a template that we show to the user.

Reviewer 2: Linkebreaks are useful so that you don't scroll too much when we read in non preview mode.

> to us when rendering a template that we show to the user.

Reviewer: Well, if you two insist. For me, who uses a really wide monitor, I prefer to choose where my paragraphs end, instead of having them predefined in the markdown. On my monitor the markdown file only covers half the file that's open 😸

```python
from src.in_memory_storage import InMemoryStorage

database = InMemoryStorage()

@app.route('/upload_word', methods=['POST'])
def upload_word():
    secret_word = request.form['secretWord']
    database.add_word(secret_word)  # save word to the db
    flash("Uploaded word " + repr(secret_word))
    flash("All available words: " + repr(database.get_all_words()))  # display all words saved so far
    return redirect('/')
```

Run the app, try uploading words again. Verify that now you get more informative messages,
and that the server now remembers all the words you've uploaded so far.

> Caveat: Because we store words in memory, this storage is alive as long as the server itself
> is alive. If you shut down the server, its memory is released, and all saved words will be forgotten.

> Caveat: storing words in memory also doesn't play well with running Flask in DEBUG mode.
> Flask in DEBUG mode restarts the server every time you've changed something in the code,
> which again means losing all the saved words.

### Part 3. The Game

#### Intro

The game of guessing words will go as follows:

- the user visits a `/game` page;

> - our server chooses a secret word the user has to guess;

Reviewer: ```suggestion
- our server randomly chooses a secret word the user has to guess;
```

- the user sends their guesses through a form until they get it right;
- if they get it right, they are congratulated and redirected back to the main page.

Note that until the user wins the game, we need to remember the secret word that they have to guess.

We did something similar when we were storing all uploaded words in the database. But this case is different:
if multiple players visit our server, we want them to play with the same, common set of words, but their

> games shouldn't interfere. Each player should play their own game, they can't all share the secret word

Reviewer: `Each player should play their own game, they can't all share the secret word
to guess like the share the pool of possible words.` <- I don't follow

Reviewer: Looks good

to guess like the share the pool of possible words.

Here we'll make use a new concept - a "session". A session makes it possible to remember information 
from one request to another, and every visitor of our website has their own session. We can
save the word to guess in the session, and it will stay the same throughout the game, until we decide
to change it ourselves.

> More about sessions - in [flask docs](flashttps://flask.palletsprojects.com/en/2.2.x/api/#sessions)

#### 1. Enable flask sessions

> Flask will create sessions only after we configure it some secret key.

Author: How about: 
```suggestion
Flask will create sessions only after we configure it some secret key.

> This is a security measure - flask stores session data in the user's browser as a cookie 
cryptographically signed with this secret key. Without a secret key, there would be 
no signature, and without a signature, hackers could tamper with the session data.

```

> Flask will create sessions only after we configure it some secret key.

Reviewer: Great

To make the game work, we saved the game context - the secret word - in the flask session.

Flask stores sessions as signed cookies in the browser, and the users, if they want, can 
find this cookie and read the data in the session!

This is a security breach - the users can cheat and always guess the secret word from the first try.
Let's try to make it harder to do for them!

#### 1. Demonstrate cheating

Start the app, upload some words, and navigate to the `/game` page. 
The server at this point should choose a word for you to guess and save it to the session.

Open the cookies in your browser.

Chrome:

`Menu > More tools > Developer Tools > Application > Storage > Cookies`

Firefox:

`Menu > More tools > Web Developer Tools > Storage > Cookies`

Safari:

- [Enable dev tools](https://support.apple.com/guide/safari/use-the-developer-tools-in-the-develop-menu-sfri20948/mac)
- `Develop > Show Web Inspector > Storage > Cookies`

Edge:

`Menu > More tools > Developer Tools > Application > Storage > Cookies`

Find a cookie with name "session" and a long value starting with "ey". It contains your encoded session data.

Copy this value to https://www.base64decode.org/, and decode it. 

> In the decoded data, you should see the secret word that you can use to guess th word from the first try!

Reviewer 2: ```suggestion
In the decoded data, you should see the secret word that you can use to guess the word from the first try!
```

        flash("You can't guess words without starting the game first!")
        return redirect('/game')

    secret_word = database.get_word_by_index(session['secret_word_id'])

    if request.form['guessed_word'] == secret_word:
        flash(Markup("You guessed right! Good job! The secret word was <b>%s</b>" % secret_word))
        del session['secret_word_id']
        return redirect('/')

    flash("You didn't guess right! Try again!")
    return redirect('/game')
```

Run the service, play the game, verify that it works.
Decode the session cookie again, verify that now it contains just the index instead of the secret word itself. 

### Part 5. Recap

We've created a real mini-game. We learned how to:

- upload data from an HTML page to the server: using forms and POST requests;
- save data in memory to share it between all users of the server as long as the server is alive;
- save data to user sessions to share it between requests of _the same user_;
- read data from the session in the browser to demonstrate that you never should store secrets in the Flask session. 

Now you can think how you can further improve this game! For example:

> - How to add hints to the game? I.e. a user would upload a word "red" with a hint "color of a tomato",

Reviewer 2: What if we inverse the first two examples so that the easiest is first?

Reviewer 2: Other ideas: Do not add a word if it already there

Reviewer 2: Other idea: keep all words in the session and allow to win the game if any word from the list is chosen. Constraint: how many words/letters can we store in a session?

Author: Good ideas!

> - How to add hints to the game? I.e. a user would upload a word "red" with a hint "color of a tomato",
>   and on the game page you'd display the hint.
> - How to make the game more forgiving to spelling? I.e. when a secret word is "red", also accept a guess "Red".
> - How to let user know how close they are to the answer for them to not guess blindly? I.e. we can consider
>   a guess "rid" very close to "red" because they share many common letters. 
> - How to restrict cheating even more? Word indexes in the session are more secure than the words  
>   themselves, but eventually the player will learn what index corresponds to which word.
> 

Author: ```suggestion
- How to make the game more forgiving about spelling? I.e. when a secret word is "red", also accept the guess "Red".
- How would you ensure the same word can't be added to the list of words more than once?
- How to let the user know how close they are to the answer for them to not guess blindly? I.e. we can consider
  a guess "rid" very close to "red" because they share many common letters. 
- How to add hints to the game? I.e. a user would upload a word "red" with a hint "color of a tomato",
  and on the game page you'd display the hint.
- How would you display the list of the wrong guesses the user has made so far in the current game?
- How to restrict cheating even more? Word indexes in the session are more secure than the words  
  themselves, but eventually the player will learn what index corresponds to which word.
```

        <label for="secretId">Secret</label>
        <input type="text" class="form-control" id="secretId" name="secretWord" aria-describedby="secretHelp" placeholder="Secret Word">
        <small id="secretHelp" class="form-text text-muted"></small>
    </div>

    <button type="submit" class="btn btn-primary">Submit</button>
</form>
```

#### 2. Serve this HTML page on HTTP endpoint

To 'serve' this page to users, add a new endpoint in `api.py`:

```python
@app.route('/words', methods=['GET'])
def upload_word_page():
    return render_template('words.html')
```

Now start the app, and verify that you can see this form on the endpoint `/words`.

> Most likely, the app will start on localhost:5000. In this case you can use this link to open the page:
> [http://localhost:5000/words](http://locahost:5000/words)

When you type in a word in the form and click "submit", your browser will try to
send it to the endpoint `/upload_word`. This is the endpoint we specified in the
`action` attribute of the form.

Because this endpoint doesn't exist yet, in the browser you'll see a "Not Found" page,
and in the logs of the running application you'll see the following entry:

> ```

Author: ```suggestion
```txt
```

# In memory storage

## Goal

Implement an interactive web app where you can:
- upload words to the game dictionary;
- play a game of guessing a word from the dictionary.

## Tutorial

> In this tutorial we will assume that you've completed the previous tutorial
and already have a python project with `app.py`, `src/api.py` and `src/templates`.

### Part 1. Uploading words

#### Intro

In the previous stage, we have implemented a so-called "static" website, where
you can get pages from the server, but their content is always the same.

Now it's time to **upload** data to the server! In our case, upload words to the dictionary.

To do this, we'll need:
1. a text field on the web page with a button "submit" (this will send data to the server)
2. an endpoint on the server that would accept this data.

#### 1. Create HTML page for word upload

Let's create a web page that will contain a text field for the word to upload and a button `submit`.

> Create a new file `upload_word.html` in the folder `templates`.

Author: ```suggestion
Create a new file `words.html` in the folder `templates`.
```

# In memory storage

## Goal

Implement an interactive web app where you can:
- upload words to the game dictionary;
- play a game of guessing a word from the dictionary.

## Tutorial

> In this tutorial we will assume that you've completed the previous tutorial
and already have a python project with `app.py`, `src/api.py` and `src/templates`.

### Part 1. Uploading words

#### Intro

In the previous stage, we have implemented a so-called "static" website, where
you can get pages from the server, but their content is always the same.

Now it's time to **upload** data to the server! In our case, upload words to the dictionary.

To do this, we'll need:
1. a text field on the web page with a button "submit" (this will send data to the server)
2. an endpoint on the server that would accept this data.

#### 1. Create HTML page for word upload

Let's create a web page that will contain a text field for the word to upload and a button `submit`.

Create a new file `upload_word.html` in the folder `templates`:

```shell

> touch templates/upload_word.html

Reviewer: We covered creating files in earlier chapters and stopped giving direct instructions on this. Do you want to keep this in?

Author: True, removed!
